engine/src/core/graphics/rendering/buffers/vertex_buffer_layout.h

```diff
@@ -0,0 +1,114 @@
+//
+// Created by blinku on 24.07.19.
+//
+
+#pragma once
+
+#include <cstdint>
+#include <vector>
+#include <string>
+#include <algorithm>
+#include "utils/types.h"
+#include "debug/bald_assert.h"
+#include "glad/glad.h"
+
+namespace Bald::Graphics {
+
+    constexpr uint32_t CalculateFloatOffset(uint32_t x) { return x * static_cast<uint32_t>(Bald::Utils::BuiltInTypeSize::Float); }
+    constexpr uint32_t CalculateIntOffset(uint32_t x) { return x * static_cast<uint32_t>(Bald::Utils::BuiltInTypeSize::Int); }
+
+    enum class ShaderBuiltInType : uint32_t {
+        Unknown = 0,
+        Float = GL_FLOAT,
+        Int = GL_INT
+    };
+
+    // TODO: Examine differences in int vs float sizes in bytes on GPU
+    enum class ShaderBuiltInTypeSize : uint32_t {
+        Unknown = 0,
+        Vec1 = CalculateFloatOffset(1),
+        Vec2 = CalculateFloatOffset(2),
+        Vec3 = CalculateFloatOffset(3),
+        Vec4 = CalculateFloatOffset(4)
+    };
+
+    class VertexBufferLayoutElement {
+    public:
+        VertexBufferLayoutElement() = delete;
+
+        VertexBufferLayoutElement(uint8_t shaderLayoutLocation, ShaderBuiltInType type, ShaderBuiltInTypeSize typeSize,
+                                  std::string name,
```

Comment:
We want to std::move the string  -> "With move semantics added to the language and the standard library updated with move constructors added for many types it is now interesting to take an argument directly by value, instead of by const-reference, and then copy. This check allows the compiler to take care of choosing the best way to construct the copy."
sos: https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html